Parametric control types

tttc3 opened the pull request:
Pull request that closes #359.

Core Changes:
-  Change `_ControlTerm` to `AbstractControlTerm[_Control]`;
- Allow type hinting of the form `ControlTerm[AbstractBrownianPath]`. Provided demonstration in `_solvers/milstein.py`. Will all relevant solvers want updating?

Additional Changes:
- Bumped ruff and pyright pre-commit versions;
- Corrected typos in `_solvers/milstein.py`.

Known Issues:
- In `test/test_term.py:133`  the following type error is thrown (not sure why):
```bash
test/test_term.py:133:46 - error: Argument of type "(t0: Unknown, t1: Unknown) -> tuple[Array]" 
cannot be assigned to parameter "control" of type "_Control@ControlTerm" in function "__init__"
    Type "(t0: Unknown, t1: Unknown) -> tuple[Array]" cannot be assigned to type "AbstractPath"
      "function" is incompatible with "AbstractPath" (reportArgumentType)
```
- In `diffrax/_term.py:266` the following type error is being suppressed (as I can't find a nice way to avoid it):
```bash
diffrax/_term.py:266:14 - error: Cannot assign member "control" for type "AbstractControlTerm[_Control@AbstractControlTerm]*"
    Expression of type "AbstractPath | _CallableToPath" cannot be assigned to member "control" of class "AbstractControlTerm[_Control@AbstractControlTerm]"
      Member "__set__" is unknown
      Type "AbstractPath | _CallableToPath" cannot be assigned to type "_Control@AbstractControlTerm" (reportAttributeAccessIssue)
```

patrick-kidger commented:
I think this looks excellent!

> Allow type hinting of the form ControlTerm[AbstractBrownianPath]. Provided demonstration in _solvers/milstein.py. Will all relevant solvers want updating?

So for this, I'm now thinking that what we really need is a way to indicate the type of the control directly in `AbstractTerm`, not just `AbstractControlTerm`. The fact that we had `_ControlTerm` was only ever just an implementation detail.
Put another way, from a user perspective, then ideally they should only need to know about `AbstractTerm` and its concrete subclasses; I don't want to introduce a whole abstract hierarchy!

It's a bit odd, but would something like this work do you think?
```python
class AbstractTerm(eqx.Module, Generic[_Control]):
    ...  # doesn't actually use _Control

class _AbstractControlTerm(AbstractTerm[_Control]):
    control: _Control
    ...
```
(possibly with a `pyright: ignore` to prevent any complaints about an unused generic parameter?)

I'm not completely sold on the design here, so happy to hear alternate suggestions.

> In test/test_term.py:133 the following type error is thrown (not sure why):

Ah, this is due to a quirk of `__init__` methods and inheritance. This takes a small amount of explanation.

Dataclasses will create a new `__init__` method on subclasses, even if an `__init__` method already exists in the parent class. (I.e. dataclasses do not inherit `__init__` methods.) So in this case pyright thinks that is what is happening, and that `ControlTerm.__init__` is being created just by looking at the fields... including `control: _Control`. (And not `control: Union[AbstractPath, Callable]`.

In practice we've found that this is a bit awkard ergonomically, so Equinox deliberately diverges from standard dataclass behaviour here, by just inheriting `__init__` methods as normal if one is provided. (If you're curious, Equinox has precisely two divergences from dataclasses: the one above, and the fact that you can write `self.foo = bar` inside `__init__` methods, rather than requiring the `object.__setattr__(self, "foo", bar)` required with regular frozen dataclasses.)

> In diffrax/_term.py:266 the following type error is being suppressed (as I can't find a nice way to avoid it):

I've bumped into versions of this one before. It's an annoying case with generics + fields + `__init__` methods that I've not found a nice way to resolve. I think adding a `pyright: ignore` annotation is indeed appropriate here.

tttc3 commented:
Thank you for the information, and explanations. They will save me from many future headaches.

Regarding the design of the type hinting for `AbstractTerm`. An alternative approach would be to type hint the output type of `contr`. For example, with the new `UnsafeBrownianPath`, you could do something like:
```python
class AbstractTerm(eqx.Module, Generic[_Control]):
    def contr(self, t0:RealScalarLike, t1:RealScalarLike) -> _Control:
        ...

# Restrict to control which evaluate to provide a LevyArea.
term: AbstractTerm[LevyArea]
# This would be valid
term = ControlTerm(UnsafeBrownianPath(..., levy_area="space-time"))
# This would be invalid.
term = ControlTerm(UnsafeBrownianPath(...))
```
This approach is more permissive than hinting on the paths themselves. It is also closer to my original problem, where I don't really care what path the user provides, only that the control evaluates to a certain shape.

The downside is that you might need to define new types for distinguishing between different controls.

patrick-kidger commented:
Ah, that's a nice alternative solution. Indeed it does mean that there's no way to say something like 'this solver only accepts Brownian motions', since our existing Brownian paths just return normal JAX arrays. Another possible difficulty is that I'm not sure how easy this will be for either runtime or static type checkers to pick up on.

However on the plus side, it's something which really does make sense for `AbstractTerm` itself -- and arguably a point of Diffrax is that you *shouldn't* need to distinguish between different types of control, with the our term system deliberately erasing any difference in types there. So I think on balance I like your suggestion better - let's go with that!

I think one change that we would want to introduce is an `AbstractLevyVal`, so that we can distinguish between the different kinds of Levy area that we might want. The goal I have in mind here is that if we're going to do parameterisation by control, then we should be able to replace this (from #344):

https://github.com/patrick-kidger/diffrax/blob/59778bdc285aa6ddbc8068c9344175fb06bb71fe/diffrax/_solver/srk.py#L285-L301

with something more principled.

tttc3 commented:
Implemented so far:
- Generic typing of `AbstractTerm.contr`
- Generic typing of `AbstractPath.evaluate` and `AbstractPath.derivative`.

Remaining:
- Abstraction of `LevyVal`.

Known Issues:
- test/test_term.py:135 from the original comment.
- Unfortunately it doesn't appear that any of mypy, pyright or beartype can spot the type violation in the example `test/test_term.py:67`.  I've tested with `reveal_type` and the Generic term remains as Unknown.

patrick-kidger commented:
> Unfortunately it doesn't appear that any of mypy, pyright or beartype can spot the type violation in the example test/test_term.py:67.

It's not obvious to me what the error is supposed to be here? `ControlTwo()` has its return type annotated as `LevyVal`, which is consistent with the `term_typed: diffrax.ControlTerm[diffrax.LevyVal]` annotation.

For what it's worth, when it comes to tackling the fact that our Brownian motions switch their return type based on an `__init__` argument, I think I've managed to figure out how to get pyright to check that (by adding some dummy annotations for `__new__`):
```python
from typing import cast, Generic, Literal, overload, TypeVar, TYPE_CHECKING, Union

T = TypeVar("T")

class Foo(Generic[T]):
    def meth(self) -> T:
        ...

class Bar(Foo[T]):
    if TYPE_CHECKING:
        @overload
        def __new__(cls, arg: Literal["use_int"]) -> "Bar[int]":
            ...

        @overload
        def __new__(cls, arg: Literal["use_str"]) -> "Bar[str]":
            ...

        def __new__(cls, arg: Literal["use_int", "use_str"]):
            if arg == "use_int":
                return cast(Bar[int], None)
            elif arg == "use_str":
                return cast(Bar[str], None)
            else:
                assert False

    def __init__(self, arg: Literal["use_int", "use_str"]):
        self.arg = arg

    def meth(self) -> T:
        if self.arg == "use_int":
            return 2  # pyright: ignore
        elif self.arg == "use_str":
            return "hi"  # pyright: ignore
        else:
            assert False

def f(x: Foo[int]):
    pass

f(Bar("use_str"))  # this produces an error with pyright!
```

tttc3 commented:
So I've managed to avoid the error in `test/test_term.py:136` by utilising a property that calls `_callable_to_path` on the `_control` attribute. 

I've also simplified the type checking tests and added a placeholder for beartype runtime type checking, which can be enabled if https://github.com/beartype/beartype/issues/238 is implemented.  Without this feature there will be no way to perform checking of any control typed as "ControlTerm[PyTree[...]]", as static checking evaluates such hints to `Unknown`.
That said, the hints will still be useful as documentation.

I like your trick for the init argument dependant hints!

Is there any reason why `use_levy` and `levy_area` both need to exist If we had types like these:

```python
class BrownianIncrement(eqx.Module):
    dt: ...
    W: ...

class SpaceTimeLevyArea(BrownianIncrement):
    H: ...
    bar_H: ...
    K: ...
    bar_k: ...
```
Maybe not exactly like this as they don't obey concrete is final?

patrick-kidger commented on diffrax/_brownian/base.py:
```diff
@@ -8,7 +8,7 @@
 
 
-class AbstractBrownianPath(AbstractPath):
+class AbstractBrownianPath(AbstractPath[Union[PyTree[Array], LevyVal]]):
```
Note that this means that `AbstractBrownianPath` can no longer be subscripted. I think you probably mean something like
```python
_Control = TypeVar("_Control", bound=Union[PyTree[Array], LevyVal])

class AbstractBrownianPath(AbstractPath[_Control]):
```
to preserve the possibility of subscripting its subclasses.

patrick-kidger commented on diffrax/_custom_types.py:
```diff
@@ -46,6 +46,7 @@
 Y = PyTree[Shaped[ArrayLike, "?*y"], "Y"]
 VF = PyTree[Shaped[ArrayLike, "?*vf"], "VF"]
 Control = PyTree[Shaped[ArrayLike, "?*control"], "C"]
+_Control = TypeVar("_Control", bound=Control)
```
On which note, I'd suggest not putting `_Control` here, but just copying it into every file it's needed. This (a) means that the `bound` can be refined, and (b) I generally try not to import `TypeVar`s, but instead treat them as part of the implementation of whatever it is they're subcripting.

(Also (c) the convention in Diffrax is that having an underscore at the start of a name indicates that it is private to the *file* it is in, not to the overall package.)

patrick-kidger commented on diffrax/_solver/milstein.py:
```diff
@@ -42,13 +43,15 @@ class StratonovichMilstein(AbstractStratonovichSolver):
 
-    term_structure: ClassVar = MultiTerm[tuple[ODETerm, AbstractTerm]]
+    term_structure: ClassVar = MultiTerm[
+        tuple[ODETerm, AbstractTerm[AbstractBrownianPath]]
```
I think this needs updating to use the return type, not the type of the control path itself.

(Which in this case I think means we can just keep the original unparameterised
```python
term_structure: ClassVar = MultiTerm[tuple[ODETerm, AbstractTerm]]
```
)

patrick-kidger commented on diffrax/_term.py:
```diff
@@ -219,45 +219,55 @@ def _mul(v):
+class _AbstractControlTerm(AbstractTerm[_Control]):
     vector_field: Callable[[RealScalarLike, Y, Args], VF]
-    control: Union[AbstractPath, Callable] = eqx.field(converter=_callable_to_path)
+    _control: Union[
+        AbstractPath[_Control], Callable[[RealScalarLike, RealScalarLike], _Control]
+    ]
+
+    @property
+    def control(self) -> AbstractPath[_Control]:
+        return self._callable_to_path(self.control)
+
+    @staticmethod
+    def _callable_to_path(
+        x: Union[
+            AbstractPath[_Control], Callable[[RealScalarLike, RealScalarLike], _Control]
+        ],
+    ) -> AbstractPath[_Control]:
+        if isinstance(x, AbstractPath):
+            return x
+        else:
+            return _CallableToPath(x)
```
To be honest, I'd probably sacrifice accuracy in the name of brevity here, and just do
```python
control: Union[AbstractPath[_Control], Callable[[RealScalarLike, RealScalarLike], _Control]] = eqx.field(converter=_callable_to_path)
```

This is a slightly-too-broad annotation for the field itself, of course, but I think it's more readable overall.

patrick-kidger commented on test/test_term.py:
```diff
@@ -42,6 +43,27 @@ def derivative(self, t, left=True):
 
+    # `# type: ignore` is used for contrapositive static type checking as per:
+    # https://github.com/microsoft/pyright/discussions/2411#discussioncomment-2028001
+    term_static: diffrax.ControlTerm[Array] = term  # noqa: F841
+    term_static_contra: diffrax.ControlTerm[diffrax.LevyVal] = term  # type: ignore # noqa: F841
```
Can we drop the `noqa` if the variable is named `_`?

patrick-kidger commented:
Oh for completeness -- regarding the `BrownianIncrement` and `SpaceTimeLevyArea` types, this is how you'd write them in a concrete-means-final way:
```python
class AbstractBrownianReturn(eqx.Module):
    dt: AbstractVar[...]
    W: AbstractVar[...]

class BrownianIncrement(AbstractBrownianReturn):
    dt: ...
    W: ...

class SpaceTimeLevyArea(AbstractBrownianReturn):
    dt: ...
    W: ...
    H: ...
    bar_H: ...
    K: ...
    bar_k: ...
```
if need be you could also introduce an `AbstractLevyArea` that sits between `SpaceTimeLevyArea` and `AbstractBrownianReturn` in the type hierarchy, if you wanted to additionally have another kind of `FooLevyArea` type.

(Or you could just keep just `AbstractBrownianReturn`, and then just use an ad-hoc supertype like `Union[SpaceTimeLevyArea, FooLevyArea]` rather than a named supertype like `AbstractLevyArea` -- this is particularly useful in cases where you don't control the type hierarchy because it's in someone else's library.)

patrick-kidger commented:
*Also also*, would it be possible to add checking for this type into the existing runtime `term_structure` check here:

https://github.com/patrick-kidger/diffrax/blob/8aafefcd081d51240ac865662a1767ca999d2455/diffrax/_integrate.py#L80

?

Probably by doing a `jax.eval_shape(term.contr, ...)` and checking the return type.

That way we can have `diffeqsolve` enforce the validity of `term_structure` (and we don't depend on beartype to do so).

patrick-kidger commented:
(let me know when you next want my input on this PR by the way)

tttc3 commented:
I think almost everything should be there now. The primary additions since the last review are:
- Corrections to directly address the review comments;
- The addition of a control check in `_integrate.py`, along with a dedicated test in `test_integrate.py`;
- The addition of `AbstractLevyArea` and the subclasses `TimeLevyArea` and `SpaceTimeLevyArea`. 
- Updated Brownian paths to expect one of the new `AbstractLevyArea` types for the `levy_area`.

`TimeLevyArea` might be better noted as `BrownianIncrement`. However, my reservation is that even if the path evaluates to a `PyTree[Array]`, semantically, we still expect this array to represent a Brownian increment of the path?

patrick-kidger commented on diffrax/_brownian/path.py:
```diff
@@ -121,25 +122,26 @@ def _evaluate_leaf(
+        dt = t1 - t0
 
-        if levy_area == "space-time":
+        if issubclass(levy_area, SpaceTimeLevyArea):
```
Nit: for clarity, I think I'd probably go for the stricter `if levy_area is SpaceTimeLevyArea` here. We shouldn't be allowing strict subclasses of `SpaceTimeLevyArea` anyway.

tttc3 replied:
I agree, the only minor issue is that when doing `levy_area is SpaceTimeLevyArea` pyright can't seem to narrow `levy_area` down to a `SpaceTimeLevyArea`. 
```python
if levy_area is SpaceTimeLevyArea:
    reveal_type(levy_area) # type[TimeLevyArea] | type[SpaceTimeLevyArea]
    levy_area(dt, w, h) # PyRight complains that `h` is in the call signature.
    SpaceTimeLevyArea(dt, w, h) # This explicit call to the initialiser is fine.
elif levy_area is TimeLevyArea:
    reveal_type(levy_area) # type[TimeLevyArea] | type[SpaceTimeLevyArea]
    levy_area(dt, w) # PyRight complains that `h` is missing from the call signature.
```
I presume this is a bug in PyRight, but I'm not sure. In any case, I can replace the `levy_area` with explicit calls to the appropriate initialiser?

patrick-kidger replied:
Ah, I'd forgotten about this. This is an intentional choice in pyright. (Although not one I agree with.)
In this case let's stick with `issubclass` then!

patrick-kidger commented on diffrax/_brownian/tree.py:
```diff
@@ -222,31 +223,31 @@ def evaluate(
-            return isinstance(obj, LevyVal)
+    ) -> Union[PyTree[Array], Union[TimeLevyArea, SpaceTimeLevyArea]]:
+        def _is_levy_area(obj):
+            return isinstance(obj, Union[TimeLevyArea, SpaceTimeLevyArea])
```
Note that `isinstance(..., Union[...])` are buggy across different Python versions. (If you're curious, I go on [a bit of a rant about this here](https://github.com/patrick-kidger/jaxtyping/issues/73#issuecomment-1493431459).) Better to use tuples instead:
```python
isinstance(obj, (TimeLevyArea, SpaceTimeLevyArea))
```

tttc3 replied:
I was not aware of this. Your post was a good read thank you!

patrick-kidger commented on diffrax/_brownian/tree.py:
```diff
@@ -222,31 +223,31 @@ def evaluate(
         # map the interval [self.t0, self.t1] onto [0,1]
         t0 = linear_rescale(self.t0, t0, self.t1)
         levy_0 = self._evaluate(t0)
+        # Should this really be returning a levy area straight away?
```
What do you mean?

tttc3 replied:
Apologies for the rogue comment, I have no idea what I was meaning either.

patrick-kidger commented on diffrax/_custom_types.py:
```diff
@@ -53,34 +52,41 @@
 
-class LevyVal(eqx.Module):
+
+class TimeLevyArea(AbstractLevyArea):
+    dt: PyTree
+    W: PyTree
```
I think I'd probably call this something like `BrownianIncrement`. (And if you object to that subclassing something called `AbstractLevyArea` then we could change the name to something like `AbstractBrownianReturn`.)

There really isn't any Levy area here so I don't think we should use that in the name.

tttc3 replied:
How about renaming `AbstractLevyArea` to `AbstractLevyVal` and then having `LevyIncrement`?

patrick-kidger replied:
I think it's really the "Levy" I'm objecting to here, which is usually taken to refer to some kind of quantity that is second-or-higher-order.

tttc3 replied:
I see, I was going to say, based on @andyElking's comment https://github.com/patrick-kidger/diffrax/pull/364#issuecomment-1938722145 `SpaceLevyArea` would be my preference, but you're far more informed on these matters then myself so I'm more than happy to do the `BrownianIncrement` if you think that is the best option.

andyElking replied:
> I see, I was going to say, based on @andyElking's comment [#364 (comment)](https://github.com/patrick-kidger/diffrax/pull/364#issuecomment-1938722145) `SpaceLevyArea` would be my preference, but you're far more informed on these matters then myself so I'm more than happy to do the `BrownianIncrement` if you think that is the best option.

I suppose calling this "Levy" might not be the most relevant anyway. It's just different terms in the signature of the path (t, W). The terms space-time and space-time-time Levy areas were introduced by James and the signature by Terry. So frankly we could easily name this after someone else...

But anyway, any control could have a signature attached to it. So maybe instead of doing this specifically for a BM we could just allow a general control to either just return its value (as with `use_levy=False`), or to also be endowed with some terms of its signature. And then we can just have a SignatureVal instead of a LevyVal?

patrick-kidger replied:
Sometimes we also find ourselves considering functions-of-the-signature-of-the-path. E.g. "full" Levy area itself is a difference of two terms.

Indeed, I think one could think about generalising what we're doing here to handle arbitrary terms in the signature. I don't think we need to worry about that here though: we can always just pursue the same strategy as we already have here, namely to have the control return some kind of special type indicating that.

Anyway, to bring this full-circle to this particular comment: I think let's call this thing `BrownianIncrement`.

patrick-kidger commented on diffrax/_custom_types.py:
```diff
@@ -53,34 +52,41 @@
-    K: Optional[PyTree]
-    bar_K: Optional[PyTree]
+    H: PyTree
+    bar_H: Optional[PyTree] = None
+    K: Optional[PyTree] = None
+    bar_K: Optional[PyTree] = None
```
@andyElking can you remind me, when are each of these values set?

Now that we explicitly distinguish at the type level whether or not the space-time Levy area is available, I'm thinking we should be able to remove the `Optional`, or move them to other subclasses of `AbstractLevyArea`?

patrick-kidger replied:
Ah, thanks!
In that case I think an internal type might make the most sense. It'd certainly help me wrap my head around the interface here :D

tttc3 replied:
Shall I remove the bar attributes from the `SpaceTimeLevyArea` in this pull then?

andyElking replied:
Yes, probably best, but they need to be kept internally in VBT (for the purposes of levy_diff). If you need, I can make an edit to VBT tomorrow so that it doesn't use LevyVal internally, and you can freely reform it elsewhere.

tttc3 replied:
Yeah that would be great thanks.

andyElking replied:
You can check PR #378

patrick-kidger commented on diffrax/_brownian/tree.py:
```diff
@@ -206,14 +211,10 @@ def sqrt_mult(z):
-            K=jtu.tree_map(sqrt_mult, x.K),
-            bar_K=None,
-        )
+        leaves, treedef = jtu.tree_flatten(x)
+        dt_normalized = [mult(leaves[0])]
+        other_normalized = jtu.tree_map(sqrt_mult, leaves[1:])
+        return jtu.tree_unflatten(treedef, dt_normalized + other_normalized)
```
So I think this is a bit dangerous, quietly assuming that `dt` is going to be the first return value. It's a little more effort, but you should be able to extract it like so:
```python
dt_normalised = mult(x.dt)
x = eqx.tree_at(lambda y: y.dt, x, None)
x = jtu.tree_map(sqrt_mult, x)
return eqx.tree_at(lambda y: y.dt, x, dt_normalised, is_leaf=lambda z: z is None)
```
which replaces the `dt` field with a `None` whilst applying the `sqrt_mult`, and then reinserts the normalised version back.

tttc3 replied:
I had originally attempted to use `eqx.partition` and `eqx.combine` with `is_leaf=lambda x: isinstance(x, AbstractLevyArea)`, and `filter_spec=lambda x: x.dt` but received TracerBooleanConversionErrors. Would you have expected this to fail? Also, would the tree_at approach actually occur in place, unlike the partition combine case?

patrick-kidger replied:
I think what's happening htere is that the `x.dt` in `filter_spec=lambda x: x.dt` is trying to be interpreted as a boolean. The intention of a filter spec is that it returns bool values to specify which half of the partition an element ends up in.
I think that could be made to work as well!

`tree_at` operates out of place. (Basically everything in JAX+Equinox does.)

patrick-kidger commented on diffrax/_brownian/tree.py:
```diff
@@ -261,27 +262,23 @@ def _evaluate_leaf(
-            bhh = None
-            bkk = None
-
+        state_key, init_key_w, init_key_la = jr.split(key, 3)
```
This was originally inside the `if self.levy_area == "space time"` branch so that we wouldn't generate an unused `init_key_la` in the common case that the output doesn't require Levy area. I think we should keep that, for efficiency.

patrick-kidger commented on diffrax/_brownian/tree.py:
```diff
@@ -474,22 +465,17 @@ def _brownian_arch(
-            bkk_st_tu = None
 
         else:
-            assert _state.bhh_s_u_su is None
```
I think we lost this assert statement in the refactor.

I'm a little antsy about the refactor here -- easiser to be sure this is doing the desired thing (e.g. like the above assert statement) if it's just the `if` statement that changes to accomodate hew new syntax.

By the way, note that the original version also used single-assignment, e.g. `bhh_stu` was assigned to precisely once, regardless of which branch was taken. In this version, it's instead assigned to as `None` at the start, and then overriden inside one of the `if` branches. Generally speaking single-assignment tends to be (I believe!) a more reliable programming style, as you don't rely on default values. (This has been a staple of FP languages for a while; more recently single-assignment is something that has been popularised by Rust.)

tttc3 replied:
Understood. I have reintroduced the assertion and ensured single assignment. If you would like me to revert all but the changes to the if statement then no worries. I couldn't find one a quick Look, but it would be nice if ruff/pyright had an option to flag multi-assignment?

patrick-kidger replied:
I think changing just the if statement is probably what I'd go for here. I'm not worried about small amounts of code duplication between the branches of the if statement (if anything I think it's easier to read when you don't have to worry about keeping the in-common code correct between both branches), and I think this will be easier for to review!

patrick-kidger commented on diffrax/_integrate.py:
```diff
@@ -73,23 +73,30 @@ class State(eqx.Module):
-            if not isinstance(term, term_cls):
-                raise ValueError
+            term_args_expected = get_args(term_cls)
+            control_compatible = True
```
Another example of single-assignment here: better to have an `else: control_compatible = True` branch, I think.

patrick-kidger commented on diffrax/_integrate.py:
```diff
@@ -73,23 +73,30 @@ class State(eqx.Module):
-                raise ValueError
+            term_args_expected = get_args(term_cls)
+            control_compatible = True
+            if term_args_expected:
```
Here's another safe-programming-style hobby-horse of mine :)
Always explicitly use booleans in `if` statements. IIUC, I think we're (currently) expecting this to be either a `None` or a length-1 tuple of the control type? In that case I would write this as:
```python
if term_args_expected is None:
    control_compatible = True
else:
    (term_arg_expected,) = term_args_expected
    ...
```

patrick-kidger commented on diffrax/_integrate.py:
```diff
@@ -73,23 +73,30 @@ class State(eqx.Module):
+            control_compatible = True
+            if term_args_expected:
+                term_cls = get_origin(term_cls)
+                t0 = getattr(term, "control.t0", -jnp.inf)
+                t1 = getattr(term, "control.t1", jnp.inf)
+                control = jax.eval_shape(term.contr, t0, t1)
```
`jax.eval_shape` will just look at the shape and dtype and ignore the rest. I think Just `jax.eval_shape(term.contr, 0.0, 0.0)` should work.

patrick-kidger commented on test/test_term.py:
```diff
@@ -42,6 +43,27 @@ def derivative(self, t, left=True):
 
+    # `# type: ignore` is used for contrapositive static type checking as per:
+    # https://github.com/microsoft/pyright/discussions/2411#discussioncomment-2028001
+    _term_static: diffrax.ControlTerm[Array] = term
+    _term_static_contra: diffrax.ControlTerm[diffrax.LevyVal] = term  # type: ignore
```
I think the variables can be just a `_` as they're unused:
```python
_: diffrax.ControlTerm[Array] = term
_: diffrax.ControlTerm[diffrax.LevyVal] = term
```

tttc3 replied:
I've just noticed that if you do something like this pyright will raise a `reportAssignmentType` on line one, as it infers the type of `_` based on the hint on line two.
```python
_: diffrax.ControlTerm[Array] = term
reveal_type(_) # diffrax.ControlTerm[diffrax.LevyVal]
_: diffrax.ControlTerm[diffrax.LevyVal] = term

# Simple fix
_: diffrax.ControlTerm[Array] = term
__: diffrax.ControlTerm[diffrax.LevyVal] = term
```

patrick-kidger reviewed (commented):
You can see my main comments are really just about a few points of either programming style (single-assignment, explicit bools in if statements) or PR style (try to minimise diffs so that they're easy for me to review!)

So basically, they're nits and I like how this looks overall :)

One question -- how do you want to tackle this alongside #370? I'm thinking we probably want the VF type to appear in the signature before the control type, e.g. `AbstractTerm[_VF, _Control]`.

tttc3 commented:
Regarding #370, I'm happy to include vector_field typing in this pull, and agree with your suggested format for`AbstractTerm[_VF, _Control]`.

andyElking commented:
Hey @tttc3 if you want a somewhat clearer overview of what the VBT is doing with Levy Areas, I have a paper in the works here:
https://www.overleaf.com/read/zqvmccrdwzsy#ccc24b

I will try to give this thread a bit more attention once I figure out how `eqxi.scan` works 😅

andyElking commented:
> `TimeLevyArea` might be better noted as `BrownianIncrement`. However, my reservation is that even if the path evaluates to a `PyTree[Array]`, semantically, we still expect this array to represent a Brownian increment of the path?

In fact the Brownian increment is a space Levy Area if anything. Time Levy area would literally just be t.

tttc3 commented:
I think, with the exception of the discussion on the LevyArea types, my recent commits should address the review comments. The only remaining unimplemented feature is the vector field type hinting for `AbstractTerm`?

tttc3 commented:
I've now added draft vector field typing support.

patrick-kidger commented on diffrax/_custom_types.py:
```diff
@@ -53,34 +52,41 @@
     - `tree_shape`: Corresponds to `outer_treedef` in `jax.tree_transpose`.
-    - `levy_area`: can be `""` or `"space-time"`, which indicates
-    which fields of the LevyVal will have values.
+    - `levy_area`: the type of `AbstractLevyReturn`
```
"`AbstractLevyReturn`" isn't used as a name anywhere else.

tttc3 replied:
Given the decision to use `BrownianIncrement`, do you want to use the `AbstractLevyReturn` or `AbstractBrownianReturn` for this type?

patrick-kidger replied:
Let's go for `AbstractBrownianReturn`!

patrick-kidger commented on diffrax/_integrate.py:
```diff
@@ -73,23 +73,57 @@ class State(eqx.Module):
 
 
-def _is_none(x):
+class _ShapeDtypeTypeCoercer(AbstractArray):
+    def __new__(cls, term_type: Any):
+        if not issubclass(term_type, AbstractArray):
+            return term_type
+
+        cls.array_type = jax.ShapeDtypeStruct
+        cls.dtypes = term_type.dtypes
+        cls.dims = term_type.dims
+        cls.index_variadic = term_type.index_variadic
+        cls.dim_str = term_type.dim_str
+        return cls
```
Haha, I see what you're doing here! FWIW the attributes on `AbstractArray` aren't public API (maybe we should add some underscores in front) so this is a bit dangerous.

To check the shape and dtype of a struct against a jaxtyping array annotation e.g. `Float[Array, "foo"]`, then I'd probably use `jax.eval_shape` to cast the struct to an array:
```python
def _assert(x):
    assert isinstance(x, Float[Array, "foo"])
jax.eval_shape(_assert, struct)
```

tttc3 replied:
Really like your solution, much better than my hack haha!

patrick-kidger commented on diffrax/_integrate.py:
```diff
@@ -73,23 +73,57 @@ class State(eqx.Module):
         else:
-            if not isinstance(term, term_cls):
-                raise ValueError
+            term_args = get_args(term_cls)
+            n_term_args = len(term_args)
+            if n_term_args >= 1:
+                _term_cls = get_origin(term_cls)
+                vector_field = jax.eval_shape(term.vf, 0.0, y, args)
+                compatible_vf_type = _ShapeDtypeTypeCoercer(term_args[0])
+                vector_field_compatible = isinstance(vector_field, compatible_vf_type)
+                if n_term_args == 2 and vector_field_compatible:
+                    control = jax.eval_shape(term.contr, 0.0, 0.0)
+                    compatible_control_type = _ShapeDtypeTypeCoercer(term_args[1])
+                    control_compatible = isinstance(control, compatible_control_type)
+                else:
+                    control_compatible = False
+                args_compatible = vector_field_compatible and control_compatible
+            else:
+                _term_cls = term_cls
+                args_compatible = True
```
This is pretty complicated :)
Can you add some comments describing how the annotations are unpacked and tackled here?

patrick-kidger commented on diffrax/_integrate.py:
```diff
@@ -73,23 +73,57 @@ class State(eqx.Module):
+            n_term_args = len(term_args)
+            if n_term_args >= 1:
+                _term_cls = get_origin(term_cls)
+                vector_field = jax.eval_shape(term.vf, 0.0, y, args)
```
This will need to be `eqx.filter_eval_shape` as `args` might include non-JAX-types.

patrick-kidger commented on diffrax/_integrate.py:
```diff
@@ -73,23 +73,57 @@ class State(eqx.Module):
+                if n_term_args == 2 and vector_field_compatible:
+                    control = jax.eval_shape(term.contr, 0.0, 0.0)
+                    compatible_control_type = _ShapeDtypeTypeCoercer(term_args[1])
+                    control_compatible = isinstance(control, compatible_control_type)
```
If the annotation is a more general type that doesn ot support `isinstance`, e.g. `tuple[int, Float[Array, "foo"]]`, then this will fail.

This calls for a runtime type checker!

As it happens, via jaxtyping we already have a transitive dependency on `typeguard==2.13.3` (this exact version), so we can assume it is present. I think add it to `pyproject.toml` to make things explicit, and then you should be able to do:
```python
def better_isinstance(x, annotation):
    @typeguard.typechecked
    def f(y: annotation):
        pass
    try:
        f(x)
    except TypeError:
        return False
    else:
        return True
```

patrick-kidger reviewed (commented):
(See also my responses to the two outstanding comments above.)

patrick-kidger commented:
Heads-up that I've just (a) merged #378 into `dev`, and (b) switched the target branch on this PR to `dev`. The conflicts now being reported are just the changes from that PR -- I think just rebase on top of `dev` and things should work.

tttc3 commented:
I've added comments to, and attempted to simplify the logic in, the `term_compatibility` check. I've also minimised the diffs where possible and extended the tests to cover the types for which `better_isinstance` is needed. Providing the new tests are sufficient, I think this is now (hopefully) feature complete.

patrick-kidger commented:
Okay, take a look at the `parametric-tweaks` branch I've just pushed. It's not 100% ready because (a) it's completely untested -- I gotta dash for now -- and (b) I want to add some tests for the tweaks I've made to the term compatbility. To explain what's going on here:
- I don't think the single-argument return from `get_args` is ever possible; Python demands that all type annotations be filled.
- However, they *can* be filled with `TypeVar`s, so I've added a check for that.
- The `y` should be tree-mapped alongside the terms. This is admittedly an entirely undocumented point that I need to make clearer somewhere at some point. (But take a look at `SemiImplicitEuler` for an example of this.)

patrick-kidger commented:
Aaaaand opened https://github.com/tttc3/diffrax/pull/1 with these changes.

patrick-kidger commented:
And merged! Gosh, that ended up being much harder than we all thought it would be, huh?

Thank you for the contribution, and great work putting this all together. I'm really happy to have this in, it's a huge improvement to the codebase. :)

tttc3 commented:
I just noticed that at some point we lost the `reportUnnecessaryTypeIgnoreComment = true` from the pyproject.toml, so the contrapositive static type checking test in `test_terms` aren't actually doing anything.

patrick-kidger commented:
Ah, agreed throughout. Can you send a PR?

dhirschfeld commented on pyproject.toml:
```diff
@@ -23,7 +23,7 @@ classifiers = [
 ]
 urls = {repository = "https://github.com/patrick-kidger/diffrax" }
-dependencies = ["jax>=0.4.23", "jaxtyping>=0.2.24", "typing_extensions>=4.5.0", "equinox>=0.11.2", "lineax>=0.0.4", "optimistix>=0.0.6"]
+dependencies = ["jax>=0.4.23", "jaxtyping>=0.2.24", "typing_extensions>=4.5.0", "typeguard==2.13.3", "equinox>=0.11.2", "lineax>=0.0.4", "optimistix>=0.0.6"]
```
Why is `typeguard` pinned to an exact version, especially one that is so old?

It's generally considered very bad packaging practice to pin to exact versions as it's ~never required and prevents getting even security patches without publishing a new version.

Also, it's ancient and won't be getting any updates - current HEAD is 4.2.1:


![image](https://github.com/patrick-kidger/diffrax/assets/881019/aedd5652-d344-4864-8644-1555088559b9)

![image](https://github.com/patrick-kidger/diffrax/assets/881019/1c751fd2-6132-413a-b575-c4900329d5b8)
